zerver/lib/import_realm.py

```diff
@@ -1995,3 +2008,54 @@ def add_users_to_system_user_groups(
         )
         for membership in usergroup_memberships
     )
+
+
+def check_migration_status(migration_status_file: str) -> None:
+    # Each server might have a different set of apps installed (e.g., Zulip Cloud has
+    # zilencer/corporate, but self-hosted might not). This function only checks for
+    # migration mismatches in apps installed on both servers.
+    local_migrations = parse_migration_status(get_migration_status())
+    exported_migrations = parse_migration_status(migration_status_file)
```

Comment:
I think you might need to strip off that first `ZULIP_VERSION` line before calling this?

I'm also now wondering if these files should just be structured JSON objects rather than text files; I don't see any obvious benefit to having the parsing/unparsing here, when it could just be `orjson.dumps/loads`, and using JSON would mean we can add additional fields freely. A structure like:

```
{
    "migrations_by_app": {
         "zerver": [...], 
         ...
    },
    "zulip_version": {version",
```

seems a lot cleaner and more extensible. We would need to use a different diffing function, but I think that might be OK?

Reply:
I've changed the code to output it as [migration_status.json](https://github.com/zulip/zulip/pull/31681/files#diff-c60963a5e58ccd293819063ca1de48e237698e0b39719d1e3f5c962924aa727a). and used `unified_diff` as the diffing function for it. This is definitely a lot cleaner since we can use TypedDict to check the format the json file.